packages/labs/analyzer/src/lib/javascript/classes.ts

```diff
@@ -94,64 +110,158 @@ const getIsReadonlyForNode = (
   return false;
 };
 
+/**
+ * Is the node a constructor assignment to `this`
+ * @example `this.foo = 'bar'`
+ */
+const isConstructorAssignmentStatement = (
+  node: ts.Node,
+  typescript: typeof ts
+): node is ts.ExpressionStatement & {
+  expression: ts.BinaryExpression & {
+    left: ts.PropertyAccessExpression;
+  };
+} =>
+  typescript.isExpressionStatement(node) &&
+  typescript.isBinaryExpression(node.expression) &&
+  node.expression.operatorToken.kind === typescript.SyntaxKind.EqualsToken &&
+  typescript.isPropertyAccessExpression(node.expression.left) &&
+  node.expression.left.expression.kind === typescript.SyntaxKind.ThisKeyword;
+
+export const deriveFieldsFromConstructorAssignments = (
+  node: ts.ConstructorDeclaration,
+  {fieldMap}: ClassMemberInfo,
+  analyzer: AnalyzerInterface
+) => {
+  // Ignore non-implementation signatures of overloaded methods by checking
+  // for `node.body`.
+  if (node.body) {
+    // TODO(bennypowers): We probably want to see if this matches what TypeScript considers a field initialization.
+    // Maybe instead of iterating through the constructor statements, we walk the body looking for any
+    // assignment expression so that we get ones inside of if statements, in parenthesized expressions, etc.
+    //
+    // Also, this doesn't cover destructuring assignment.
+    //
+    // This is ok for now because these are rare ways to "declare" a field,
+    // especially in web components where you shouldn't have constructor parameters.
+    node.body.statements.forEach((node) => {
+      if (isConstructorAssignmentStatement(node, analyzer.typescript)) {
+        const name = node.expression.left.name.getText();
+        fieldMap.set(
```

Comment:
what do we do if there's both a field _declaration_ and a constructor assignment? I think we should ignore the constructor assignment then, yeah?

Reply:
Map set handles that, right? So comment that order of operations is important?

Reply:
ok I added a guard which should handle cases like this:
```ts
class A {
  /** don't clobber me */
  a = 'a';
  constructor() {
    this.a = 'a';
  }
}
```